docs/src/design/many_differentials.md

```diff
@@ -0,0 +1,183 @@
+So beyond _natural_ differential types, we also have _structural_ differential types.
+ChainRules uses `Composite{P, <:NamedTuple}` to represent a structural differential type corresponding to primal type `P`.
+`Zygote v0.4` uses `NamedTuple`.
+Structural differentials are derived from the structure of the input.
+Either automatically, as part of the AD, or manually as part of a custom rule.
+
+Consider the structure of `DateTime`:
+```julia
+julia> dump(now())
+DateTime
+  instant: UTInstant{Millisecond}
+    periods: Millisecond
+      value: Int64 63719890305605
+```
+
+The corresponding structural differential is;
+```julia
+Composite{DateTime}(
+    instant::Composite{UTInstant{Millisecond}}(
+        periods::Composite{Millisecond}(
+                value::Int64
+        )
+    )
+)
+
+```
+
+!!! note “One must be allowed to take derivatives of integer arguments”
+    This brings up another contrast to Swift.
+    In Swift `Int` is considered non-differentiable, which is quite reasonable; it doesn’t have a very good definition of the limit of a small step (as that would be some floating/fixed point type).
+    `Int` is intrinsically discrete.
+    It is commonly used for indexing, and if one takes a gradient step, say turning `x[2]` into `x[2.1]` then that is an error.
+    However, disallowing `Int` to be used as a differential means we can not handle cases like `DateTime` having a inner field of milliseconds counted as an integer from the unix epoch; or other cases where an integer is used as a convenience for computational efficiency.
+    In the case where a custom sensitivity rule claims that there is a non-zero derivative for an `Int` argument that is being used for indexing, that code is simply wrong.
+    We can’t handle incorrect code and trying to is a path toward madness.
+    Julia, unlike Swift, is not well suited to handling rules about what you can and can’t do with particular types.
+
+So the structural differential is another type of differential.
+Since AD can only create structural differentials, unless using custom sensitivity rules; and since all custom sensitivities are only written in terms of natural differentials, as that is what is used in papers about derivatives.
+So you need to support both.
+
+## Semi-structural differentials
+
+Where there is no natural differential type for the outermost type, but there is for some of its fields, which can call this a semi-structural differential.
+
+Consider if we had a representation of a country's GDP as output by some continuous time model like a Gaussian Process, where that representation is as a sequence of `TimeSample`s.
+
+Where a `TimeSample` is a structure as follows:
+```julia
+julia> struct TimeSample
+           time::DateTime
+           value::Float64
+       end
+```
+
+We can look at its structure:
+julia> dump(TimeSample(now(), 2.6e9))
```

Comment:
need to open triple-backticks before this line

Reply:
this one's important cos it breaks rendering after this point

Reply:
```suggestion
```julia
julia> dump(TimeSample(now(), 2.6e9))
```